MA-2422: Implemented nav buttons and swipe in landscape

mdinino opened the pull request:
### Description

[MA-2422](https://openedx.atlassian.net/browse/MA-2422)

Implemented nav bar and swiping in courseware landscape mode

### Testing
- [ ] Unit, integration, acceptance tests as appropriate
 - Opened edX Demonstration course and navigated to courseware. Clicked on "Demo Course Overview" and then "Blank HTML Page".
- In portrait mode clicked on next and previous buttons several times. Ensured that worked properly.
- Swiped with finger left and right. Ensured that navigation occurred properly
- Rotated phone to landscape mode. Ensured that we are still on correct "courseware". Also ensured that nav button appeared at bottom.
- Clicked on next/previous buttons several times and ensured that it works as expected.
- Swipe with finger left and right several times and ensured that it works properly.
- Rotate back to landscape mode and repeat those tests.
- Turn the screen off/on a few times and repeat tests.

### Reviewers
If you've been tagged for review, please check your corresponding box once you've given the :+1:.
- [ ] Code review: @BenjiLee 
- [ ] Code review: @bguertin
- [ ] UX review: @lizcohen

mdinino commented:
@lizcohen Wants me to change the behavior of the nav bar so that it always says "Next" or "Previous" instead of sometimes saying "Next Unit" or "Previous Unit". I will fix that with this PR.

Otherwise she gave verbally gave me the 👍

BenjiLee commented:
FYI you have failing tests.

mdinino commented:
Yeah, I have to fix those. Woking on it now

On Fri, Jul 1, 2016 at 11:28 AM, Christopher Lee <notifications@github.com>
wrote:

> FYI you have failing tests.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/edx/edx-app-android/pull/761#issuecomment-229976346>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AR10T72kLJr0n2zwi0j73LAoQoU9P0ojks5qRTICgaJpZM4JCZsX>
> .
>



-- 
Regards,

Marc DiNino
Android Engineer
EDX

mdinino commented:
@BenjiLee and @bguertin or @miankhalid, I fixed the failing test's and addressed Liz's request. Review please?

BenjiLee commented:
👍

mdinino commented:
@bguertin just a friendly reminder to review

miankhalid commented on VideoLocker/src/main/java/org/edx/mobile/view/CourseUnitNavigationActivity.java:
```diff
@@ -180,12 +180,10 @@ private void tryToUpdateForEndOfSequential(){
             String nextUnitSubsectionId = unitList.get(curIndex + 1).getParent().getId();
```
>// fix: https://openedx.atlassian.net/browse/MA-995
// code below decides to show Next/Previous Unit name or only Next/Previous
// based on units in a subsection

These comments should also be deleted.

miankhalid commented on VideoLocker/src/main/java/org/edx/mobile/view/CourseUnitNavigationActivity.java:
```diff
@@ -228,7 +224,7 @@ private void updateDataModel(){
         EnumSet<BlockType> types =  userPrefManager.isUserPrefVideoModel() ?
                              EnumSet.of(BlockType.VIDEO) : EnumSet.allOf(BlockType.class);
-        ((CourseComponent) selectedUnit.getRoot()).fetchAllLeafComponents(leaves, types);
+        (selectedUnit.getRoot()).fetchAllLeafComponents(leaves, types);
```
Might as well get rid of the extra parenthesis surrounding `selectedUnit.getRoot()`.

miankhalid commented:
Other than a couple of nits. LGTM 👍

bguertin commented:
👍

bguertin commented on VideoLocker/src/main/java/org/edx/mobile/view/CourseUnitNavigationActivity.java:
```diff
@@ -259,12 +252,8 @@ public void onConfigurationChanged(Configuration newConfig) {
     private void updateUIForOrientation(){
         if (getResources().getConfiguration().orientation == Configuration.ORIENTATION_LANDSCAPE) {
             setActionBarVisible(false);
```
@mdinino I don't think the toolbar should be hidden either (It's not hidden on iOS). So in fact, this whole method can be removed.

bguertin commented:
When you are in full-screen mode, The Next/Prev bar is overlapping the video controls.

I think when you are in full-screen mode, we shouldn't show the prev/next bar?

lizcohen commented:
@bguertin I agree, user should leave full screen mode to navigate prev/next

mdinino commented:
@bguertin @lizcohen @BenjiLee That's whole point of this JIRA though. In landscape mode, videos are full screen, and the ticket wanted to implement next/prev in that mode.

Should we allow swipe gesture support in that mode but no buttons? Or just close this JIRA because it's intended behavior?

bguertin commented:
@mdinino I don't think rotating the device should automatically be full-screen (especially for certain devices that are primarily designed for landscape use). Also seems like an accessability problem.

I think there is supposed to be a non-fullscreen landscape view, that includes the toolbar and next/prev nav.

Full-screen should ONLY be entered if you tap the "view full-screen" button.

bguertin commented:
Note, there seems to be a bug in the non-fullscreen landscape view at the moment. https://openedx.atlassian.net/browse/MA-2583

cc @lizcohen

mdinino commented:
It seems like the "rotate for fullscreen" was intended per this screenshot
![screen shot 2016-07-11 at 11 51 47 am](https://cloud.githubusercontent.com/assets/18707535/16737142/fc4ae104-475d-11e6-9eae-5b5259843f8f.png)

mdinino commented:
cc @marcotuts

marcotuts commented:
I will follow up offline with @lizcohen on the intent behind this change. 

I would put this on hold for a bit from a dev standpoint until we sync up. 

Thanks!

BenjiLee commented:
@mdinino I have updated the acceptance criteria in the ticket.

mdinino commented:
@BenjiLee and @bguertin : I updated the implementation to reflect the new acceptance criteria and fixed the automated tests.

Can you re-review?

mdinino commented:
Also, @marcotuts Gave me a verbal 👍

BenjiLee commented:
Titlebar disappears when swiping through units. 

Repro steps:
1. Get to units
2. Swipe to next unit
3. if a unit is 1 swipe away from a video unit, the title bar disappears.

I'm guessing that this is because the code is hiding the titlebar when it see's a video unit. When using viewpager, it preloads a set number of pages. In our case, it is 3. That means if the next unit is a video, and the titlebar is supposed to hide on a video unit, it is going to try to hide the title bar. since the video unit has been loaded but not visible.

mdinino commented:
@BenjiLee Can you re-review? You can use my device to test before you go to far down the rabbit hole.

BenjiLee commented on VideoLocker/src/main/java/org/edx/mobile/view/CourseUnitNavigationActivity.java:
```diff
@@ -106,32 +108,57 @@ public void onPageScrollStateChanged(int state) {
+            pager.setCurrentItem(index - 1);
+            ret = true;
+        }
+
```
Could you remove all of these empty lines in this method and below (navigateNextComponent)

mdinino replied:
Yes, also I just merged Vfrom master (and pushed), and I lost a tad bit of
my work. Recreating now… sorry about that.

On Tue, Aug 2, 2016 at 3:33 PM, Christopher Lee <notifications@github.com>
wrote:

> In
> VideoLocker/src/main/java/org/edx/mobile/view/CourseUnitNavigationActivity.java
> <https://github.com/edx/edx-app-android/pull/761#discussion_r73221848>:
>
> >          updateUIForOrientation();
> >      }
> >
> >      @Override
> > +    public boolean navigatePreviousComponent() {
> > +        boolean ret = false;
> > +
> > +        int index = pager.getCurrentItem();
> > +        if (index > 0) {
> > +            PageViewStateCallback curView = (PageViewStateCallback) pagerAdapter.instantiateItem(pager, index);
> > +            if (curView != null)
> > +                curView.onPageDisappear();
> > +            pager.setCurrentItem(index - 1);
> > +            ret = true;
> > +        }
> > +
>
> Could you remove all of these empty lines in this method and below
> (navigateNextComponent)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/edx/edx-app-android/pull/761/files/74aa7add97dd18b9508daecea42bb12965de956f#r73221848>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AR10T-i_MEca1ci-ezCaUgxqfcvcyRsXks5qb5ukgaJpZM4JCZsX>
> .
>



-- 
Regards,

Marc DiNino
Android Engineer
EDX

mdinino commented:
@BenjiLee I took out the whitespace (and fixed my merge error)

mdinino commented:
Tests are failing in Travis, but working on my machine. I'll try squashing my commits.

BenjiLee commented on VideoLocker/src/main/java/org/edx/mobile/view/CourseUnitFragment.java:
```diff
@@ -14,6 +14,8 @@
 public abstract class CourseUnitFragment extends BaseFragment implements PageViewStateCallback, RunnableCourseComponent {
     public interface HasComponent {
         CourseComponent getComponent();
+        boolean navigateNextComponent();
```
Why are these booleans?

mdinino replied:
It returns false if there is no next/previous.

BenjiLee replied:
But why do we need to know that?

mdinino replied:
At the moment, nothing. When I first made it, it mattered. I shifted the
logic around and now nothing needs it, but I kept in the return variable.
Do you want to make it void? That’s a simple change.

On Wed, Aug 3, 2016 at 1:46 PM, Christopher Lee <notifications@github.com>
wrote:

> In VideoLocker/src/main/java/org/edx/mobile/view/CourseUnitFragment.java
> <https://github.com/edx/edx-app-android/pull/761#discussion_r73379174>:
>
> > @@ -14,6 +14,8 @@
> >  public abstract class CourseUnitFragment extends BaseFragment implements PageViewStateCallback, RunnableCourseComponent {
> >      public interface HasComponent {
> >          CourseComponent getComponent();
> > +        boolean navigateNextComponent();
>
> Let me clarify, what calls this that requires a boolean value.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/edx/edx-app-android/pull/761/files/64c1f08c0f3651fb25188c643146d7accc62968d#r73379174>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AR10T58gdPxJXsJZ53wuFxqIU7kR8yCcks5qcNPvgaJpZM4JCZsX>
> .
>



-- 
Regards,

Marc DiNino
Android Engineer
EDX

BenjiLee replied:
If nothing uses it now and there is no concrete plan or reason to use it in the future, than remove it.

mdinino replied:
Removed.

On Wed, Aug 3, 2016 at 2:17 PM, Christopher Lee <notifications@github.com>
wrote:

> In VideoLocker/src/main/java/org/edx/mobile/view/CourseUnitFragment.java
> <https://github.com/edx/edx-app-android/pull/761#discussion_r73389486>:
>
> > @@ -14,6 +14,8 @@
> >  public abstract class CourseUnitFragment extends BaseFragment implements PageViewStateCallback, RunnableCourseComponent {
> >      public interface HasComponent {
> >          CourseComponent getComponent();
> > +        boolean navigateNextComponent();
>
> If nothing uses it now and there is no concrete plan or reason to use it
> in the future, than remove it.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/edx/edx-app-android/pull/761/files/64c1f08c0f3651fb25188c643146d7accc62968d#r73389486>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AR10TxmN0NIDfbzKyC5fR3X-qPDP-EsKks5qcNtKgaJpZM4JCZsX>
> .
>



-- 
Regards,

Marc DiNino
Android Engineer
EDX

BenjiLee commented on VideoLocker/src/main/java/org/edx/mobile/view/CourseUnitVideoFragment.java:
```diff
@@ -158,16 +161,38 @@ public void onActivityCreated(Bundle savedInstanceState) {
             isLandscape = false;
         } else {
             isLandscape = true;
```
You can remove this if else. Nothing uses `isLandscape`. You can also remove variable declaration.

mdinino replied:
Will do.

BenjiLee commented on VideoLocker/src/main/java/org/edx/mobile/view/CourseUnitVideoFragment.java:
```diff
@@ -75,19 +69,26 @@
     private final Handler playHandler = new Handler();
     private View messageContainer;
 
+    private final static String HAS_NEXT_UNIT_ID = "has next unit";
```
Use underscores for stringdefs. I don't know what the official AOSP style is, but we should at least keep it consistent with our other stringdefs.

mdinino replied:
fixed

BenjiLee commented on VideoLocker/src/main/java/org/edx/mobile/view/CourseUnitVideoFragment.java:
```diff
@@ -639,43 +669,40 @@ public void onFail(Exception ex) {
-    @Override
-    public void onConfigurationChanged(Configuration newConfig) {
-        super.onConfigurationChanged(newConfig);
+    private void updateUIForOrientation(boolean isLandscape) {
```
Let's remove the boolean argument here since we can just call `getResources().getConfiguration().orientation == Configuration.ORIENTATION_LANDSCAPE`

mdinino replied:
Addressed

mdinino commented:
@BenjiLee I addressed your comments (and cleaned up some other errors too). In my last checkin, the tests failed, but it seemed like it was an odd build error rather than a test. Hopefully it will work this time.

BenjiLee commented:
Done with pass. Should be good, as long as the tests are not problematic.

bguertin commented on VideoLocker/src/main/java/org/edx/mobile/view/CourseUnitNavigationActivity.java:
```diff
@@ -263,11 +272,27 @@ public void onConfigurationChanged(Configuration newConfig) {
+        int currentItem = pager.getCurrentItem();
+
+        if (currentItem < pagerAdapter.getCount()) {
+            fragment = pagerAdapter.getItem(currentItem);
```
This instantiates a new fragment, which we then discard. That seems a pretty roundabout way to see if the selected unit is a video.

These 9 lines could be replaced with 1:
`final boolean isCurrentFragmentVideoUnit  = selectedUnit instanceof VideoBlockModel &&
                ((VideoBlockModel) selectedUnit).getData().encodedVideos.getPreferredVideoInfo() != null`

Granted, we could move that check to a method so that the logic is shared between this and the implementation of `pagerAdapter.getItem()`

bguertin commented on VideoLocker/src/main/java/org/edx/mobile/view/CourseUnitNavigationActivity.java:
```diff
@@ -263,11 +272,27 @@ public void onConfigurationChanged(Configuration newConfig) {
             setActionBarVisible(false);
             findViewById(R.id.course_unit_nav_bar).setVisibility(View.GONE);
-            pager.setEnabled(false);
+            pager.setEnabled(true);
```
We can remove the calls to `setEnabled()` since the pager is always enabled now.

bguertin commented on VideoLocker/src/main/java/org/edx/mobile/view/adapters/CourseUnitPagerAdapter.java:
```diff
@@ -55,7 +55,7 @@ public Fragment getItem(int pos) {
         if (unit instanceof VideoBlockModel &&
                 ((VideoBlockModel) unit).getData().encodedVideos.getPreferredVideoInfo() != null) {
-            unitFragment = CourseUnitVideoFragment.newInstance((VideoBlockModel) unit);
+            unitFragment = CourseUnitVideoFragment.newInstance((VideoBlockModel) unit, (pos<unitList.size()), (pos>0));
```
nit-pick: please auto-format

bguertin commented:
Otherwise, looks good 👍

bguertin commented:
This was merged as separate PR

mdinino commented:
@bguertin, Stuart Young from Testing CI wanted me to keep this open so he could investigate it.

mdinino commented:
Actually, I he can still investigate this PR if it's closed
